File: app/src/main/java/com/example/desiregallery/viewmodels/PostListViewModel.java
package com.example.desiregallery.viewmodels;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;
import android.util.Log;
import com.example.desiregallery.MainApplication;
import com.example.desiregallery.models.Post;
import com.example.desiregallery.network.DGNetwork;
import com.example.desiregallery.network.IDGApi;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

import java.util.List;

public class PostListViewModel extends ViewModel {
    private static final String TAG = PostListViewModel.class.getSimpleName();

    private MutableLiveData<List<Post>> posts;

    public LiveData<List<Post>> getPosts() {
        if (posts == null) {
            posts = new MutableLiveData<>();
            loadPosts();
        }
        return posts;
    }

    private void loadPosts() {
        IDGApi api = DGNetwork.getService();
        api.getPosts().enqueue(new Callback<List<Post>>() {
            @Override
            public void onResponse(Call<List<Post>> call, Response<List<Post>> response) {
                if (response.isSuccessful() && response.body() != null) {
                    posts.setValue(response.body());
                    Log.i(TAG, "Posts have been loaded");
                }
            }

            @Override
            public void onFailure(Call<List<Post>> call, Throwable t) {
                Log.e(TAG, "Unable to load posts");
                t.printStackTrace();
            }
        });
    }

}

File: app/src/main/java/com/example/desiregallery/viewmodels/PostListViewModel.kt
package com.example.desiregallery.viewmodels

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.example.desiregallery.utils.logError
import com.example.desiregallery.utils.logInfo
import com.example.desiregallery.utils.logWarning
import com.example.desiregallery.data.models.Post
import com.example.desiregallery.data.network.baseService
import com.example.desiregallery.data.network.query.OrderDirection
import com.example.desiregallery.data.network.query.QueryRequest
import com.example.desiregallery.data.network.queryService
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.*

class PostListViewModel : ViewModel() {
    val posts = MutableLiveData<MutableList<Post>>()
    val isLoading = MutableLiveData<Boolean>().apply { value = false }

    var currentPage = 0
    var isLastPage = false
    val pageSize: Int
        get() = PAGE_SIZE

    init {
        loadPosts()
    }

    fun loadPosts() {
        val query = buildQuery(currentPage+1)
        isLoading.value = true
        queryService.getPosts(query).enqueue(object: Callback<List<Post>> {

            override fun onResponse(call: Call<List<Post>>, response: Response<List<Post>>) {
                if (!response.isSuccessful)
                    logWarning(
                        TAG,
                        "Failed to load posts. Response received with code ${response.code()}: ${response.message()}"
                    )
                else if (response.body() == null)
                    logWarning(
                        TAG,
                        "Failed to load posts. Response received an empty body"
                    )
                else if (response.body()!!.isEmpty()) {
                    isLastPage = true
                    logInfo(TAG, "Last page reached")
                }
                else {
                    posts.value = response.body() as MutableList<Post>
                    logInfo(TAG, "Posts have been loaded")
                    currentPage++
                }
                isLoading.value = false
            }

            override fun onFailure(call: Call<List<Post>>, t: Throwable) {
                logError(TAG, "Unable to load posts: ${t.message}")
                isLoading.value = false
            }
        })
    }

    private fun buildQuery(page: Int): QueryRequest {
        val offset = (PAGE_SIZE * (page-1)).toLong()
        return QueryRequest()
            .from("posts")
            .orderBy("timestamp", OrderDirection.DESCENDING)
            .limit(PAGE_SIZE)
            .offset(offset)
    }

    fun addPost(post: Post) {
        val currPosts = posts.value as MutableList
        currPosts.add(0, post.apply { timestamp = Date().time })
        posts.value = currPosts
        baseService.createPost(post.id, post).enqueue(object: Callback<Post> {
            override fun onFailure(call: Call<Post>, t: Throwable) {
                logError(
                    TAG,
                    "Unable to create post ${post.id}: ${t.message}"
                )
            }

            override fun onResponse(call: Call<Post>, response: Response<Post>) {
                if (!response.isSuccessful)
                    logError(
                        TAG,
                        "Unable to create post ${post.id}: received response with code ${response.code()}"
                    )
                else
                    logInfo(
                        TAG,
                        "Post ${post.id} successfully created"
                    )
            }
        })
    }

    companion object {
        private val TAG = PostListViewModel::class.java.simpleName
        const val PAGE_SIZE = 10
    }
}
